src/sb16.js

```diff
@@ -0,0 +1,1607 @@
+{
+    switch(register)
+    {
+        case 0:
+            if(bits & 0x80)
+            {
+                // IQR Reset
+            }
+            else
+            {
+                // Timer masks and on/off
+            }
+            break;
+        case 1:
+            // Four-operator enable
+            break;
+    }
+});
+
+register_fm_write([0x05], function(bits, register, address)
+{
+    if(register === 0)
+    {
+        // No registers documented here.
+        this.fm_default_write(bits, register, address);
+        return;
+    }
+    // OPL3 Mode Enable
+});
+
+register_fm_write([0x08], function(bits, register, address)
+{
+    // Composite sine wave on/off
+    // Note select (keyboard split selection method)
+});
+
+register_fm_write(between(0x20, 0x35), function(bits, register, address)
+{
+    var operator = get_fm_operator(register, address - 0x20);
+    // Tremolo
+    // Vibrato
+    // Sustain
+    // KSR Envelope Scaling
+    // Frequency Multiplication Factor
+});
+
+register_fm_write(between(0x40, 0x55), function(bits, register, address)
+{
+    var operator = get_fm_operator(register, address - 0x40);
+    // Key Scale Level
+    // Output Level
+});
+
+register_fm_write(between(0x60, 0x75), function(bits, register, address)
+{
+    var operator = get_fm_operator(register, address - 0x60);
+    // Attack Rate
+    // Decay Rate
+});
+
+register_fm_write(between(0x80, 0x95), function(bits, register, address)
+{
+    var operator = get_fm_operator(register, address - 0x80);
+    // Sustain Level
+    // Release Rate
+});
+
+register_fm_write(between(0xA0, 0xA8), function(bits, register, address)
+{
+    var channel = address - 0xA0;
+    // Frequency Number (Lower 8 bits)
+});
+
+register_fm_write(between(0xB0, 0xB8), function(bits, register, address)
+{
+    // Key-On
+    // Block Number
+    // Frequency Number (Higher 2 bits)
+});
+
+register_fm_write([0xBD], function(bits, register, address)
+{
+    // Tremelo Depth
+    // Vibrato Depth
+    // Percussion Mode
+    // Bass Drum Key-On
+    // Snare Drum Key-On
+    // Tom-Tom Key-On
+    // Cymbal Key-On
+    // Hi-Hat Key-On
+});
+
+register_fm_write(between(0xC0, 0xC8), function(bits, register, address)
+{
+    // Right Speaker Enable
+    // Left Speaker Enable
+    // Feedback Modulation Factor
+    // Synthesis Type
+});
+
+register_fm_write(between(0xE0, 0xF5), function(bits, register, address)
+{
+    var operator = get_fm_operator(register, address - 0xE0);
+    // Waveform Select
+});
+
+
+
+//
+// FM behaviours
+//
+
+
+
+SB16.prototype.fm_update_waveforms = function()
+{
+}
+
+
+
+//
+// General behaviours
+//
+
+
+
+SB16.prototype.sampling_rate_change = function(rate)
+{
+    this.sampling_rate = rate;
+}
+
+SB16.prototype.get_channel_count = function()
+{
+    return this.dsp_stereo? 2 : 1;
+}
+
+SB16.prototype.dma_transfer_size_set = function()
+{
+    this.dma_sample_count = 1
+        + (this.write_buffer.shift() << 0)
```

Comment:
Super-nitpick, but the operator should be at the end of the previous line to make it more obvious that the line continues. Also, this code is not easily verifiable to be correct in the presence of automatic semicolon insertion.